tests/assembly/sanitizer/kcfi/emit-arity-indicator.rs

```diff
@@ -0,0 +1,27 @@
+// Verifies that KCFI arity indicator is emitted.
+//
+//@ add-core-stubs
+//@ revisions: x86_64
+//@ assembly-output: emit-asm
+//@[x86_64] compile-flags: --target x86_64-unknown-linux-gnu -C llvm-args=-x86-asm-syntax=intel -Ctarget-feature=-crt-static -Zsanitizer=kcfi -Zsanitizer-kcfi-arity
+//@[x86_64] needs-llvm-components: x86
+//@ min-llvm-version: 20.1.0
+
+#![crate_type = "lib"]
+
+pub fn foo(f: fn(i32) -> i32, arg: i32) -> i32 {
+    // CHECK-LABEL: __cfi__{{.*}}3foo{{.*}}:
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  nop
+    // CHECK-NEXT:  mov eax, 653723426
```

Comment:
@ojeda @scottconstable I think I've it mostly done, but I suspect this test isn't right. If I understood it correctly, according to the table inhttps://github.com/llvm/llvm-project/pull/121070#issue-2758314082, this operation should use EDX(?) because callers would pass two parameters before calling foo it in RDI and RSI(?). 

(Note that this isn't actually arity that is being indicated. This function's arity is three (3): two parameters, and a return value.)

Reply:
Hmm... I am not sure what you mean. Arity is the number of parameters, without the return value, so it should use 2, thus it should use EDX.

From a quick test with Clang, that is the case. I also get the same result if I take `rustc`'s LLVM IR and add the arity module flag.

Are you seeing a different result?

By the way, why don't we test a simpler function, e.g. one that just sums two numbers? i.e. the intention here is to notice the difference between the disabled and enabled cases.

Reply:
Yes, the return value is not included. You're right. What I was noting there, if I understood it correctly, is that "arity" is being defined there in that PR and table at the ABI level with certain things being excluded such as parameters passed on the stack, which would cause it the "arity" defined there to diverge from the actual arity of a function, but that is not relevant here, I was just pointing it out again as I did when it was initially proposed.

Let me try a simpler pair of functions and get back to you.

Reply:
I'm still getting EAX used with a new test using a simpler pair of functions with different arity. My LLVM version is

```
rcvalle@rcvalle:~/rust$ build/x86_64-unknown-linux-gnu/stage1/bin/rustc --verbose --version
rustc 1.87.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.87.0-dev
LLVM version: 20.1.0
```

Which should include it, right?

Reply:
> Which should include it, right?

I don't think so, upstream LLVM 20.1.0 does not contain it.

Reply:
Oh, that is the reason then. What you think we should do now? Wait for it to be tagged upstream and come in to Rust's LLVM, or implement it without the LLVM version checks and the assembly test for now (and I add them later)? I guess it'd depend how urgent you need this feature on the Linux kernel side and we can make this call.

Reply:
So I wanted this one landed in nightly mostly so that the kernel people could test it easily by downloading the nightly, but if `rustc`'s LLVM does not have the patch anyway, that does not help much.

I also don't want to give too much trouble to `rustc` maintainers applying a patch to their LLVM either just for this.

I will probably build a test toolchain for Peter with this PR applied for the time being.

Reply:
SG! Let me know if there is anything I can help with in the meantime. If needed, I can also remove the checks and test and get it merged, and add these later. Just let me know!

Reply:
If you look at the llvm-main tag, you can see that we regularly condition things things to not-yet-released LLVM (e.g. LLVM 21 in this case) so that they start working as soon as released, or if `rustc` is built against a pre-release LLVM

Reply:
Will take a look. Thank you! Should we condition this on LLVM 21? Or will there be any minor version release that will have this feature?

Reply:
I would condition this on LLVM 21 - My understanding of current LLVM version numbers are:

* "Version X.0.Y" - Development version that will eventually become version X. Y is mostly meaningless, features may come and go.
* "Version X.1.Y" - LLVM Version X, stable/released. Patchlevel Y updated to indicate bugfix backports.

So I'd just condition on the major version, e.g. >= 21.0, so that dev versions start working immediately.

Reply:
Thank you for all the information, @maurer! These were very helpful! I changed all the conditionals to check for LLVM version 21 instead then.